File: tasks-api-java/src/main/java/br/com/darlan/tasks/util/TaskUtil.java
package br.com.darlan.tasks.util;

import br.com.darlan.tasks.controller.TaskController;
import br.com.darlan.tasks.dto.TaskDTO;
import br.com.darlan.tasks.model.Task;
import org.springframework.hateoas.Link;
import org.springframework.stereotype.Component;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;

import java.util.List;

@Component
public class TaskUtil {
    public TaskDTO entityToDTO(Task entity) {
        return new TaskDTO(entity.getIdTask(), entity.getTaskName(), entity.getTaskDescription(),
                entity.getUniqueExecution(), entity.getStartDate(), entity.getStartTime(),
                entity.getEndDate(), entity.getEndTime(), entity.getAddress(), entity.getHost());
    }

    public Task dtoToEntity(TaskDTO dto) {
        return Task.builder()
                .idTask(dto.getIdTask())
                .taskName(dto.getTaskName())
                .taskDescription(dto.getTaskDescription())
                .uniqueExecution(dto.getUniqueExecution())
                .startDate(dto.getStartDate())
                .startTime(dto.getStartTime())
                .endDate(dto.getEndDate())
                .endTime(dto.getEndTime())
                .address(dto.getAddress())
                .host(dto.getHost())
                .build();
    }

    public List<TaskDTO> entityListToDTOList(List<Task> entities) {
        return entities.stream().map(this::entityToDTO).toList();
    }

    public List<Task> dtoListToEntityList(List<TaskDTO> dtos) {
        return dtos.stream().map(this::dtoToEntity).toList();
    }

    public List<TaskDTO> addSelfLink(List<TaskDTO> tasks) {
        return tasks.stream().peek(this::addSelfLink).toList();
    }

    public TaskDTO addSelfLink(TaskDTO t) {
        return t.add(linkTo(methodOn(TaskController.class).findById(t.getIdTask())).withSelfRel());
    }
}

File: kotlin/tasks-api-kotlin/src/main/kotlin/br/com/darlan/tasks/utils/TaskUtil.kt
package br.com.darlan.tasks.utils

import br.com.darlan.tasks.controller.TaskController
import br.com.darlan.tasks.dto.TaskDTO
import br.com.darlan.tasks.model.Task
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn
import org.springframework.stereotype.Component

@Component
class TaskUtil {
    fun entityToDto(entity: Task): TaskDTO {
        return TaskDTO(entity.idTask, entity.taskName, entity.taskDescription,
            entity.uniqueExecution, entity.startDate, entity.startTime, entity.endDate,
            entity.endTime, entity.address, entity.host)
    }

    fun dtoToEntity(dto: TaskDTO): Task {
        return Task(dto.idTask, dto.taskName, dto.taskDescription,
            dto.uniqueExecution, dto.startDate, dto.startTime, dto.endDate,
            dto.endTime, dto.address, dto.host)
    }

    fun entityListToDTOList(entities: List<Task>): List<TaskDTO> {
        return entities.stream().map { e -> entityToDto(e) }
            .toList()
    }

    fun dtoListToEntityList(dtos: List<TaskDTO>): List<Task> {
        return dtos.stream().map { d -> dtoToEntity(d) }
            .toList()
    }

    fun addSelfLink(dtos: List<TaskDTO>): List<TaskDTO> {
        return dtos.stream().peek { d -> run {
            d.add(linkTo(methodOn(TaskController::class.java).findById(d.idTask)).withSelfRel())
        } }.toList()
    }
}